Sources/NIOSSL/NIOSSLHandler.swift

```diff
@@ -914,42 +928,68 @@ extension NIOSSLHandler {
         // These are some annoying variables we use to persist state across invocations of
         // our closures. A better version of this code might be able to simplify this somewhat.
         var promises: [EventLoopPromise<Void>] = []
-        var didWrite = false
 
         do {
             var invokeCloseOutput = false
-            bufferedActionsLoop: while bufferedActions.hasMark {
-                let element = bufferedActions.first!
-                switch element {
-                case .write(let bufferedWrite):
-                    var data = bufferedWrite.data
-                    let writeSuccessful = try self._encodeSingleWrite(buf: &data)
-                    if writeSuccessful {
-                        didWrite = true
-                        if let promise = bufferedWrite.promise { promises.append(promise) }
-                        _ = bufferedActions.removeFirst()
+            var bufferedActionsLoopCount = 0
+            bufferedActionsLoop: while self.bufferedActions.hasMark {
+                bufferedActionsLoopCount += 1
+                var didWrite = false
+
+                writeLoop: while self.bufferedActions.hasMark {
+                    let element = self.bufferedActions.first!
+                    switch element {
+                    case .write(let bufferedWrite):
+                        var data = bufferedWrite.data
+                        let writeSuccessful = try self._encodeSingleWrite(buf: &data)
+                        if writeSuccessful {
+                            didWrite = true
+                            if let promise = bufferedWrite.promise { promises.append(promise) }
+                            _ = self.bufferedActions.removeFirst()
+                        } else if didWrite {
+                            // The write into BoringSSL unsuccessful. This happens when BoringSSL
+                            // throws a 'want read' or 'want write' error.
+                            //
+                            // As a write was previously successful it's assumed that the error is
+                            // 'wants write'. Break the write loop so any data is written to the
+                            // network before resuming.
+                            break writeLoop
+                        } else {
+                            // No write was successful in this loop, so assume the error
+                            // was 'wants read'.
```

Comment:
> Relatedly, I'm not sure calling `doDecodeData` in the loop here is a good idea. I don't think we should have any pending buffered data, we usually clear it on `readComplete`.
> 
> Is the expectation here that we're actually in the handshake? If that's true, I wonder if we should rethink the state management here and check for that state, rather than indirectly try to handle want read and want write.

Being in the implicit handshake is one way we can get to this point. I'm not sure what other situations would get us into this state. With that said though, I think the additional state checking prevents us from getting into the implicit handshake.

I didn't feel entirely comfortable about these changes (especially the `else` branch) but leaving the potential spin loop also didn't sit right.

Given this code – aside from the known path into the spin loop – has worked fine for years, perhaps we should back out these changes and just keep the iteration limit?

Reply:
I think the iteration limit can work, combined with checking our state and delaying any flushing until after we think the handshake is done.

Reply:
> I think the iteration limit can work

To clarify, do you mean just the iteration limit as-is? Or without the current `else` branch? Or something else?

> combined with checking our state and delaying any flushing until after we think the handshake is done.

Okay I think we have our bases covered then, we only call `doUnbufferActions(context:)` from three different places:

1. `channelRead(context:data:)` if the state is `active` or `outputClosed`
2. `completeHandshake(context:)`
3. `flush(context:)` if state is one of `.active`, `.unwrapping`, `.closing`, `.unwrapped`, `.inputClosed`, `.outputClosed`, `.closed`

So we should only be unbuffering actions and flushing if we're in an appropriate state.

Reply:
> Or without the current else branch?

Without the current `else` branch.